### PR TITLE
Reordered RX-TX pairs to be consistent

### PR DESCRIPTION
**Pull request by SergioGasquez**
## Thank you for your contribution!

We appreciate the time and effort you've put into this pull request.
To help us review it efficiently, please ensure you've gone through the following checklist:

### Submission Checklist 📝
- [x] I have updated existing examples or added new ones (if applicable).
- [x] I have used `cargo xtask fmt-packages` command to ensure that all changed code is formatted correctly.
- [x] My changes were added to the [`CHANGELOG.md`](https://github.com/esp-rs/esp-hal/blob/main/esp-hal/CHANGELOG.md) in the **_proper_** section.
- [x] I have added necessary changes to user code to the [Migration Guide](https://github.com/esp-rs/esp-hal/blob/main/esp-hal/MIGRATING-0.21.md).
- [x] My changes are in accordance to the [esp-rs API guidelines](https://github.com/esp-rs/esp-hal/blob/main/documentation/API-GUIDELINES.md)

#### Extra:
- [x] I have read the [CONTRIBUTING.md guide](https://github.com/esp-rs/esp-hal/blob/main/documentation/CONTRIBUTING.md) and followed its instructions.

#### Description
Reordered RX-TX pairs in all APIs to be consistent

#### Testing
HIL tests

### CODE REVIEW

**Review by bugadani — APPROVED** [review]
Maybe doing this internally is a bit excessive, but thanks for doing it anyway :)

**Review by JurajSadel — APPROVED** [review]
Thanks!

File: esp-hal/MIGRATING-0.20.md [hunk 1]
```diff
@@ -35,3 +35,25 @@ Instead of manually grabbing peripherals and setting up clocks, you should now c
      // ...
  }
 ```
+## RX/TX Order
+
+Previously, or API was pretty inconsitent with the RX/TX ordering, and different peripherals had different order. Now, all
```
Comment (bugadani) [inline esp-hal/MIGRATING-0.20.md:40]:
```suggestion
Previously, our API was pretty inconsitent with the RX/TX ordering, and different peripherals had different order. Now, all
```

File: esp-hal/MIGRATING-0.20.md [hunk 2]
```diff
@@ -35,3 +35,25 @@ Instead of manually grabbing peripherals and setting up clocks, you should now c
...
+...
+
+- let transfer = spi
+-    .dma_transfer(dma_tx_buf, dma_rx_buf)
+-    .map_err(|e| e.0)
+-    .unwrap();
++ let transfer = spi
++    .dma_transfer(dma_rx_buf, dma_tx_buf)
++    .map_err(|e| e.0)
++    .unwrap();
```
Comment (bugadani) [inline esp-hal/MIGRATING-0.20.md:58]:
```suggestion
  let transfer = spi
-    .dma_transfer(dma_tx_buf, dma_rx_buf)
+    .dma_transfer(dma_rx_buf, dma_tx_buf)
     .map_err(|e| e.0)
     .unwrap();
```

**Review by bugadani — APPROVED** [review]
Thanks :)

**Comment by SergioGasquez** [comment]
Regarding the HIL failures, I was able to run the test suite locally for C6 and S3 (having pins 2-3 and 5-6 connected) and H2 (connections have some issues atm, @playfulFence is working on a fix)

File: hil-test/tests/spi_full_duplex_dma.rs [hunk 3]
```diff
@@ -79,67 +79,67 @@ mod tests {
...
+        let (_, (dma_rx_buf, dma_tx_buf)) = transfer.wait();
         assert_eq!(dma_tx_buf.as_slice(), dma_rx_buf.as_slice());
     }
 
     #[test]
     #[timeout(3)]
     fn test_asymmetric_dma_transfer(ctx: Context) {
-        let (tx_buffer, tx_descriptors, rx_buffer, rx_descriptors) = dma_buffers!(4, 2);
-        let mut dma_tx_buf = DmaTxBuf::new(tx_descriptors, tx_buffer).unwrap();
+        let (rx_buffer, rx_descriptors, tx_buffer, tx_descriptors) = dma_buffers!(2, 4);
```
Comment (SergioGasquez) [inline hil-test/tests/spi_full_duplex_dma.rs:100]: For some reason, [S2 HIL](https://github.com/esp-rs/esp-hal/actions/runs/10717649673/job/29718015468) is failing this test, I can also reproduce it locally.  Not sure why it only happens in S2 and when using different buffer sizes (using `dma_buffers!(4,4)` passes the test)

Comment (JurajSadel) [reply hil-test/tests/spi_full_duplex_dma.rs:100]: @SergioGasquez I tried to run it on ESP32 and test fails there as well :/

Comment (jessebraham) [reply hil-test/tests/spi_full_duplex_dma.rs:100]: I recall we were having issues with that test not long ago, but perhaps I'm mistaken.

Comment (SergioGasquez) [reply hil-test/tests/spi_full_duplex_dma.rs:100]: It was disabled for s2 before https://github.com/esp-rs/esp-hal/pull/1990, but after that it was working fine (and still is,on main). Looks like something is wrong in pdma when using different size buffers, but I cant see what.

Comment (SergioGasquez) [reply hil-test/tests/spi_full_duplex_dma.rs:100]: Looks like the other tests are leaving the initial state for the test wrong. Using the following example works fine:

```rust
//! SPI Full Duplex DMA Issue Repro
//% CHIPS: esp32 esp32c2 esp32c3 esp32c6 esp32h2 esp32s2 esp32s3

#![no_std]
#![no_main]

use esp_backtrace as _;
use esp_hal::{
    dma::{Dma, DmaPriority, DmaRxBuf, DmaTxBuf},
    dma_buffers,
    gpio::Io,
    prelude::*,
    spi::{master::Spi, SpiMode},
};
use esp_println::println;

#[entry]
fn main() -> ! {
    let peripherals = esp_hal::init(esp_hal::Config::default());

    let io = Io::new(peripherals.GPIO, peripherals.IO_MUX);
    let sclk = io.pins.gpio0;
    let miso = io.pins.gpio9;
    let mosi = io.pins.gpio10;
    let cs = io.pins.gpio8;

    let dma = Dma::new(peripherals.DMA);

    let dma_channel = dma.spi2channel;

    let spi = Spi::new(peripherals.SPI2, 100.kHz(), SpiMode::Mode0)
        .with_pins(Some(sclk), Some(mosi), Some(miso), Some(cs))
        .with_dma(dma_channel.configure(false, DmaPriority::Priority0));

    let (rx_buffer, rx_descriptors, tx_buffer, tx_descriptors) = dma_buffers!(2, 4);
    let dma_rx_buf = DmaRxBuf::new(rx_descriptors, rx_buffer).unwrap();
    let mut dma_tx_buf = DmaTxBuf::new(tx_descriptors, tx_buffer).unwrap();

    dma_tx_buf.fill(&[0xde, 0xad, 0xbe, 0xef]);

    let transfer = spi
        .dma_transfer(dma_rx_buf, dma_tx_buf)
        .map_err(|e| e.0)
        .unwrap();
    let (_, (dma_rx_buf, dma_tx_buf)) = transfer.wait();
    assert_eq!(dma_tx_buf.as_slice()[0..1], dma_rx_buf.as_slice()[0..1]);
    println!("TX: {:x?}", dma_tx_buf.as_slice());
    println!("RX: {:x?}", dma_rx_buf.as_slice());
    loop {}
}
```
Also, if we comment all the tests in `spi_full_duplex_dma` but `test_asymmetric_dma_transfer` (the one that fails), the tests passes.

Comment (bugadani) [reply hil-test/tests/spi_full_duplex_dma.rs:100]: That is especially sad because the MCU is supposed to be fully reset between tests.

Comment (JurajSadel) [reply hil-test/tests/spi_full_duplex_dma.rs:101]: Just to double-check, did you try to run this example multiple times?

**Pull request merged**